src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -129,7 +129,6 @@ public void to(MutationComposer composer) {
 
   @VisibleForTesting
   void toSerializable() {
-    // TODO: scan might cause anomalies since it might cause phantoms ?
```

Comment:
Phantoms have not gone. We should keep this comment.

Initial state:

|`key`|`sub_key`|`value`|
|-----|---------|-------|
|`foo`|`t0`|1|
|`foo`|`t3`|1|

- T1: Scan where `'t0' <= sub_key and sub_key < 't3'` and store the sum of `value` into `sum1`.
- T2: Scan where `'t1' <= sub_key and sub_key < 't4'` and store the sum of `value` into `sum2`.
- T1: Put `sum1` into key ("foo", "t1").
- T2: Put `sum2` into key ("foo", "t2").
- T1: Commit. Checks `t0` and `t1` then succeeds.
- T2: Commit. Checks `t3` and `t2` then succeeds.

Final state:

|`key`|`sub_key`|`value`|
|-----|---------|-------|
|`foo`|`t0`|1|
|`foo`|`t1`|1|
|`foo`|`t2`|1|
|`foo`|`t3`|1|

Neither T1 → T2 nor T2 → T1 results in this state.

Reply:
Great 👍 
I think scan and write in a transaction could cause phantom/write skew in the current implementation (without something like predicate locking, which is a little hard to implement in the current architecture), so it probably needs to be like this.
```
              if (e.getValue().isPresent() && !writeSet.isEmpty()) {
```

Reply:
Probably,
R U W ⊇ S => OK (Serializable) 

R: a set of keys in read set by get
W: a set of keys in write set by put
S: a set of keys in scan ( ∞ for scan on empty records because keys could be anything)

Reply:
Counter example (initial state are same to the previous example):

- T1: Scan where `'t0' <= sub_key and sub_key < 't3'` and store the sum of `value` into `sum1`.
- T2: Scan where `'t1' <= sub_key and sub_key < 't4'` and store the sum of `value` into `sum2`.
- T1: Put `sum1` into key ("foo", "t1") **and ("foo", "t0")**.
- T2: Put `sum2` into key ("foo", "t2") **and ("foo", "t3")**.
- T1: Commit. Checks `t0` and `t1` then succeeds.
- T2: Commit. Checks `t3` and `t2` then succeeds.

Reply:
T1 and T2 are reading non-existing records. so S is infinite.